src/lib/classify.ml

```diff
@@ -381,78 +381,44 @@ module GaussianNaiveBayes(Data: Continuous_encoded_data_intf)
 
   end
 
-module LogisticRegression(Data: Continuous_encoded_data_intf)
-  : (Classifier_intf with type feature = Data.feature
-                     and type clas = Data.clas
-                     and type spec = unit)
+type log_reg_spec =
+  { lambda    : float
+  ; tolerance : float
+  }
 
-  = struct
+module LrCommon(Data: Continuous_encoded_data_intf) = struct
 
   open Lacaml_D
 
-  module Log_reg = struct
-
-    (* Code modified from
-      http://math.umons.ac.be/anum/fr/software/OCaml/Logistic_Regression/ *)
-
-    (* [logistic_grad_n_eval] returns the value of the function to maximize
-       and store its gradient in [g]. *)
-    let logistic_grad_n_eval ~lambda x y =
-      (fun w g ->
-        let s = ref 0. in
-        ignore(copy ~y:g w);                   (* g ← w *)
-        scal (-. lambda) g;                    (* g = -λ w *)
-        for i = 0 to Array.length x - 1 do
-          let yi = y.(i) in
-          let e  = exp(-. yi *. dot w x.(i)) in
-          s := !s +. log1p e;
-          axpy x.(i) ~alpha:(yi *. e /. (1. +. e)) g;
-        done;
-        -. !s -. 0.5 *. lambda *. dot w w)
-
-    (* TODO. expose lambda as possible [spec]. *)
-    let log_reg ?(lambda=0.1) x y =
-      let w = Vec.make0 (Vec.dim x.(0)) in
-      ignore(Lbfgs.F.max (*~print:(Lbfgs.Every 10) *)
-              (logistic_grad_n_eval ~lambda x y) w);
-      w
-
-  end
-
   type feature = Data.feature
   type clas = Data.clas
 
   type samples = (clas * feature) list
 
-  type t =
-    { weights : vec
-    ; classes : (clas * float) list
-    }
+  type spec = log_reg_spec
+  let default = { lambda = 1e-4
+                ; tolerance = 1e4
+                }
 
   let safe_encoding =
     to_safe_encoding_size_checked "Continuous_encoded_data_intf"
       Data.size Data.encoding
 
-  let proba w x y = Float.(1. / (1. + exp(-. y * dot w x)))
-
-  let eval lr feature =
-    let a = safe_encoding feature in
-    let m = Vec.init (Data.size + 1) (function | 1 -> 1.0 | i -> a.(i - 2)) in
-    List.map (fun (c,c_i) -> c, proba lr.weights m c_i) lr.classes
+  let copy1 arr = Array.init (Data.size + 1) (function | 0 -> 1. | i -> arr.(i - 1))
 
-  type spec = unit
-  let default = ()
-
-  let estimate ?(spec=default) ?(classes=[]) data =
+  (* map classes to [1;2 ... 3], convert features to matrix and run Softmax *)
```

Comment:
Why 1-based, instead of 0-based? And why is class_bound a required argument--is that something a client will be using often? (No suggestions for changes, just trying to understand.)